### ISSUE TITLE
Usage of conditional options

### ISSUE THREAD

**Issue by sixten-hilborn** [conan-io/conan#889]
Hi,

If I try to set a conditional option (an option that may or may not exist, depending on other options and settings etc) of a package from my conanfile, I'll get an error if this option does not exist, see lasote/conan-boost#37 for example.

In some cases it's easy to not set that option in my conanfile if the condition is simple, however a package could have a very complex relation between that option and other parameters, which may even change between releases of that package etc.

It's also a good thing that conan raises errors when an invalid option is specified, to prevent other mistakes like misspelling.
One solution is to add support for marking options as conditional, allowing them to be passed to a package without an error even though it is missing/disabled.

**Comment by memsharded** [comment]
What would be an use case for a "conditional" option? The idea is that when something changes, it is much better to raise an error, that let it pass inadvertently, because this later causes lots of pains and hours of debugging, because the library you are linking with happens to be built without the option you are passing to it.

The case you are describing  in the  lasote/conan-boost#37 example is very relevant. It is very important that you know that the option fPIC has disappeared. Otherwise, you might be debugging strange link errors, and the issue is that boost is not built with fPIC anymore. Does it make sense?

The #878 PR is not really related to this, it is mostly an internal refactor, decoupling options from settings internal representation. Also, removes the existence of nested options, nobody are using them and they could be very problematic if used, so we prefer to remove them now.

**Comment by sixten-hilborn** [comment]
I agree with your concerns.
For the Boost example, the issue is not that fPIC has been removed, but that it has been removed **if** Visual Studio is the compiler.

I think a change should be in the packages, not in their consumption. So instead of this example from Boost package:
```
if self.settings.compiler == "Visual Studio":
    self.options.remove("fPIC")
```
we could have a `disable` function:
```
if self.settings.compiler == "Visual Studio":
    self.options.disable("fPIC")
```

That would remove the fPIC option for VS, but not cause an error if using VS and specifying the fPIC option. A warning or info message `Boost:fPIC option is unused` could still be shown.

**Comment by memsharded** [comment]
Uhmmm, I don't know what to think...

In my opinion:

```bash
$ conan install .... -s compiler="Visual Studio" ... -o Boost:fPIC=True
```

Should raise an error. From my experience, warnings are highly unnoticed in the conan output.

**Comment by amerry** [comment]
I've also run into this.

Your example of running on the command line is very reasonable - if you specify an option that won't have any effect, producing an error can save pain down the line, and it's easy enough to adjust the command on the fly.

However, consider a project's dependencies specified in a `conanfile.txt` or `conanfile.py`. The one I'm looking at right now is libusb from bincrafters, which has two options that may or may not exist: enable_udev only exists on Linux, and fPIC doesn't exist with Visual Studio.

If I want to change these from the defaults (for everyone who builds the project - there is a canonical way to build this code), I can't use the nice, clean `conanfile.txt` format, and I have to encode mirror logic for when these options are available in the `conanfile.py`. While this is possible, it is inconvenient.

As an alternative to changing the upstream projects, having a way of specifying an option in such a way that it won't complain if it gets ignored would be nice. So you could specify something like:

    [requires]
    libusb/1.0.21@bincrafters/stable

    [options]
    libusb:enable_udev?=False

where the "?=" means "apply if the option exists, but don't error if it doesn't".

**Comment by memsharded** [comment]
Have you tried?

```
[options]
libusb*:enable_udev=False
```
This won't fail if the option doesn't exist.

**Comment by amerry** [comment]
No, that hadn't occurred to me, even though I was aware of the wildcard notation (just doing "*:enable_udev" seemed like it would make the file hard to follow, as it wouldn't be clear what package was supposed to be affected). Thanks!

**Comment by memsharded** [comment]
Good. Let us know if it works, if everything is good, then please close the issue. Thanks!

**Comment by amerry** [comment]
It works for what I want - I'm not the original reporter, though.

**Comment by memsharded** [comment]
@amerry oh, that is right, my mistake.

I still think that 

```bash
$ conan install .... -s compiler="Visual Studio" ... -o Boost:fPIC=True
```
Should raise an error, not be a warning. If the consumers want more complex logic to enable, disable options, then I am not willing to make conanfile.txt more complex using a custom DSL in conanfile.txt (like ``windows:x86=>lib:option=True``), and the way to go is definitely use a ``conanfile.py`` with the python syntax. 

I think that this issue was open for a long time without strong evidence of more users support, so I am closing it, but please feel free to comment or reopen if necessary. Thanks!
